### ISSUE TITLE
URL containing quote should 404, not 500

### ISSUE THREAD

**Issue by henare** [openaustralia/publicwhip#1065]
`[They Vote For You/production] URI::InvalidURIError: bad URI(is not URI?): /members/representatives/wide_bay/warren_truss/policies/8'"`

## Backtrace

[View full backtrace and more info at honeybadger.io](https://app.honeybadger.io/projects/37542/faults/14517291)

**Comment by novemberkilo** [comment]
Is this still an issue? 

https://theyvoteforyou.org.au/people/representatives/wide_bay/warren_truss/policies/8'" does not give me a 404 or a 500.

(I can't see the project on honeybadger so don't know if we've seen recent instances of the problem).

**Comment by henare** [comment]
Well spotted :eyes: I cannot reproduce the above but here's one from Honeybadger:

https://theyvoteforyou.org.au/mp.php?dmp=15%27%22&house=representatives&mpc=Hunter&mpn=Joel_Fitzgibbon

**Comment by henare** [comment]
Stacktrace from Honeybadger as requested:

```

        common.rb 176 split(...)
        /home/deploy/.rvm/rubies/ruby-2.0.0-p353/lib/ruby/2.0.0/uri/common.rb:176:in `split'

    174 
    175       else
    176         raise InvalidURIError, "bad URI(is not URI?): #{uri}"
    177       end

        common.rb 211 parse(...)
        /home/deploy/.rvm/rubies/ruby-2.0.0-p353/lib/ruby/2.0.0/uri/common.rb:211:in `parse'
        common.rb 747 parse(...)
        /home/deploy/.rvm/rubies/ruby-2.0.0-p353/lib/ruby/2.0.0/uri/common.rb:747:in `parse'
        redirection.rb 28 call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/routing/redirection.rb:28:in `call'
        router.rb 73 block in call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/journey/router.rb:73:in `block in call'
        router.rb 59 each(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/journey/router.rb:59:in `each'
        router.rb 59 call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/journey/router.rb:59:in `call'
        route_set.rb 690 call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/routing/route_set.rb:690:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        agent_hooks.rb 26 traced_call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/rack/agent_hooks.rb:26:in `traced_call'
        middleware_tracing.rb 55 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:55:in `call'
        browser_monitoring.rb 23 traced_call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/rack/browser_monitoring.rb:23:in `traced_call'
        middleware_tracing.rb 55 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:55:in `call'
        manager.rb 35 block in call(...)
        [GEM_ROOT]/gems/warden-1.2.3/lib/warden/manager.rb:35:in `block in call'
        manager.rb 34 catch(...)
        [GEM_ROOT]/gems/warden-1.2.3/lib/warden/manager.rb:34:in `catch'
        manager.rb 34 call(...)
        [GEM_ROOT]/gems/warden-1.2.3/lib/warden/manager.rb:34:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        etag.rb 23 call(...)
        [GEM_ROOT]/gems/rack-1.5.5/lib/rack/etag.rb:23:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        conditionalget.rb 25 call(...)
        [GEM_ROOT]/gems/rack-1.5.5/lib/rack/conditionalget.rb:25:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        head.rb 11 call(...)
        [GEM_ROOT]/gems/rack-1.5.5/lib/rack/head.rb:11:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        params_parser.rb 27 call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/middleware/params_parser.rb:27:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        flash.rb 254 call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/middleware/flash.rb:254:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        id.rb 225 context(...)
        [GEM_ROOT]/gems/rack-1.5.5/lib/rack/session/abstract/id.rb:225:in `context'
        id.rb 220 call(...)
        [GEM_ROOT]/gems/rack-1.5.5/lib/rack/session/abstract/id.rb:220:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        cookies.rb 562 call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/middleware/cookies.rb:562:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        query_cache.rb 36 call(...)
        [GEM_ROOT]/gems/activerecord-4.1.14.1/lib/active_record/query_cache.rb:36:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        connection_pool.rb 621 call(...)
        [GEM_ROOT]/gems/activerecord-4.1.14.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:621:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        callbacks.rb 29 block in call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
        callbacks.rb 82 run_callbacks(...)
        [GEM_ROOT]/gems/activesupport-4.1.14.1/lib/active_support/callbacks.rb:82:in `run_callbacks'
        callbacks.rb 27 call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/middleware/callbacks.rb:27:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        remote_ip.rb 76 call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/middleware/remote_ip.rb:76:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        debug_exceptions.rb 17 call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        show_exceptions.rb 30 call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        logger.rb 38 call_app(...)
        [GEM_ROOT]/gems/railties-4.1.14.1/lib/rails/rack/logger.rb:38:in `call_app'
        logger.rb 20 block in call(...)
        [GEM_ROOT]/gems/railties-4.1.14.1/lib/rails/rack/logger.rb:20:in `block in call'
        tagged_logging.rb 68 block in tagged(...)
        [GEM_ROOT]/gems/activesupport-4.1.14.1/lib/active_support/tagged_logging.rb:68:in `block in tagged'
        tagged_logging.rb 26 tagged(...)
        [GEM_ROOT]/gems/activesupport-4.1.14.1/lib/active_support/tagged_logging.rb:26:in `tagged'
        tagged_logging.rb 68 tagged(...)
        [GEM_ROOT]/gems/activesupport-4.1.14.1/lib/active_support/tagged_logging.rb:68:in `tagged'
        logger.rb 20 call(...)
        [GEM_ROOT]/gems/railties-4.1.14.1/lib/rails/rack/logger.rb:20:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        request_id.rb 21 call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/middleware/request_id.rb:21:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        methodoverride.rb 21 call(...)
        [GEM_ROOT]/gems/rack-1.5.5/lib/rack/methodoverride.rb:21:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        runtime.rb 17 call(...)
        [GEM_ROOT]/gems/rack-1.5.5/lib/rack/runtime.rb:17:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        local_cache_middleware.rb 26 call(...)
        [GEM_ROOT]/gems/activesupport-4.1.14.1/lib/active_support/cache/strategy/local_cache_middleware.rb:26:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        sendfile.rb 112 call(...)
        [GEM_ROOT]/gems/rack-1.5.5/lib/rack/sendfile.rb:112:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        ssl.rb 24 call(...)
        [GEM_ROOT]/gems/actionpack-4.1.14.1/lib/action_dispatch/middleware/ssl.rb:24:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        engine.rb 514 call(...)
        [GEM_ROOT]/gems/railties-4.1.14.1/lib/rails/engine.rb:514:in `call'
        application.rb 144 call(...)
        [GEM_ROOT]/gems/railties-4.1.14.1/lib/rails/application.rb:144:in `call'
        middleware_tracing.rb 57 call(...)
        [GEM_ROOT]/gems/newrelic_rpm-3.9.7.266/lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
        thread_handler_extension.rb 77 process_request(...)
        /home/deploy/.rvm/gems/ruby-1.8.7-p370/gems/passenger-4.0.10/lib/phusion_passenger/rack/thread_handler_extension.rb:77:in `process_request'
        thread_handler.rb 140 accept_and_process_next_request(...)
        /home/deploy/.rvm/gems/ruby-1.8.7-p370/gems/passenger-4.0.10/lib/phusion_passenger/request_handler/thread_handler.rb:140:in `accept_and_process_next_request'
        thread_handler.rb 108 main_loop(...)
        /home/deploy/.rvm/gems/ruby-1.8.7-p370/gems/passenger-4.0.10/lib/phusion_passenger/request_handler/thread_handler.rb:108:in `main_loop'
        request_handler.rb 441 block (3 levels) in start_threads(...)
        /home/deploy/.rvm/gems/ruby-1.8.7-p370/gems/passenger-4.0.10/lib/phusion_passenger/request_handler.rb:441:in `block (3 levels) in start_threads'
```

**Comment by novemberkilo** [comment]
tl;dr I don't think there's anything to fix here?

Why do we want to respond with a 404 instead of a 500 please?

The issue here is that the URI is invalid and so it can't be parsed. A 404 says that a particular resource cannot be found. If I understand [my response codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Response_codes) correctly, a bad/invalid URI should result in a 400 (bad request) or a 500.

In our situation, `URI.parse` is falling over in the Rails/Rack machinery and well before any app related code. So, apparently the 500 is Rails' way of handling an invalid URI.

fwiw github gives us a 500 for a mangled URI too - try: https://github.com/foo/bar/baz%$%```'''221312

**Comment by henare** [comment]
Sweet, thanks for looking into this @novemberkilo :sunglasses:
